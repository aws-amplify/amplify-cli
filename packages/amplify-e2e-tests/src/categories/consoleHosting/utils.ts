import * as fs from 'fs-extra';
import * as path from 'path';
import { HOSTING, RESOURCE, TYPE, TYPE_UNKNOWN, CATEGORIES, APPID, PROVIDER } from './constants';

export function loadTypeFromTeamProviderInfo(cwd: string, currEnv: string) {
  const teamProviderPath = path.join(cwd, 'amplify', 'team-provider-info.json');
  const content = readJsonFile(teamProviderPath);
  if (
    content &&
    content[currEnv] &&
    content[currEnv][CATEGORIES] &&
    content[currEnv][CATEGORIES][HOSTING] &&
    content[currEnv][CATEGORIES][HOSTING][RESOURCE] &&
    content[currEnv][CATEGORIES][HOSTING][RESOURCE][TYPE]
  ) {
    return content[currEnv][CATEGORIES][HOSTING][RESOURCE][TYPE];
  } else {
    return TYPE_UNKNOWN;
  }
}

export function loadAppIdFromTeamProviderInfo(cwd: string, currEnv: string) {
  const teamProviderPath = path.join(cwd, 'amplify', 'team-provider-info.json');
  const content = readJsonFile(teamProviderPath);
  console.log('content:*******');
  console.log(currEnv);
  if (content && content[currEnv] && content[currEnv][PROVIDER] && content[currEnv][PROVIDER][APPID]) {
    return content[currEnv][PROVIDER][APPID];
  } else {
    return TYPE_UNKNOWN;
  }
}

function readJsonFile(jsonFilePath: string, encoding = 'utf8') {
  return JSON.parse(fs.readFileSync(jsonFilePath, encoding));
}
