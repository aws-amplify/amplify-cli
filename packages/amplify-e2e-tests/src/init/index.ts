export {
  default as initJSProjectWithProfile,
  initProjectWithAccessKey,
  initNewEnvWithAccessKey,
  initIosProjectWithProfile,
  initAndroidProjectWithProfile,
  initNewEnvWithProfile,
} from './initProjectHelper';
export * from './amplifyPush';
export { getProjectMeta } from '../utils';
export { default as deleteProject } from './deleteProject';
