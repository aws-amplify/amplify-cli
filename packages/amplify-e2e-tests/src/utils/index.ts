import * as path from 'path';
import * as fs from 'fs-extra';
import * as rimraf from 'rimraf';
import { config } from 'dotenv';
export * from './projectMeta';
export * from './transformConfig';
export * from './awsExports';
export * from './sdk-calls';
export * from './api';

// run dotenv config to update env variable
config();

export function getCLIPath() {
  return path.join(__dirname, '..', '..', '..', 'amplify-cli', 'bin', 'amplify');
}

export function createNewProjectDir(root?: string): string {
  if (!root) {
    root = path.join(__dirname, '../../../..', `amplify-integ-${Math.round(Math.random() * 100)}-test-${Math.round(Math.random() * 1000)}`);
  }
  fs.mkdirSync(root);
  return root;
}

export function deleteProjectDir(root: string) {
  return rimraf.sync(root);
}

export function isCI(): Boolean {
  return process.env.CI ? true : false;
}

export function getEnvVars(): { ACCESS_KEY_ID: string; SECRET_ACCESS_KEY: string } {
  return { ...process.env } as { ACCESS_KEY_ID: string; SECRET_ACCESS_KEY: string };
}
