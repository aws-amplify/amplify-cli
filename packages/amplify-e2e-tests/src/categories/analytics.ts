import * as nexpect from '../utils/nexpect-modified';
import { getCLIPath, isCI } from '../utils';

export function addAnalytics(cwd: string, settings: any, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['add', 'analytics'], { cwd, stripColors: true, verbose })
      .wait('Provide your pinpoint resource name:')
      .sendline(settings.wrongName)
      .wait('Resource name should be alphanumeric')
      .sendline('\r')
      .send('\b')
      .sendline(settings.rightName)
      .wait('Apps need authorization to send analytics events. Do you want to allow guests')
      .sendline('n')
      .wait(`Successfully added resource ${settings.rightName} locally`)
      .sendEof()
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function removeAnalytics(cwd: string, settings: any, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['remove', 'analytics'], { cwd, stripColors: true, verbose })
      .wait('Choose the resource you would want to remove')
      .send('j')
      .sendline('\r')
      .wait('Are you sure you want to delete the resource?')
      .send('y')
      .sendline('\r')
      .wait('Successfully removed resource')
      .sendEof()
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}
