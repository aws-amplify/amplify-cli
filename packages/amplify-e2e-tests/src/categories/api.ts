import * as nexpect from 'nexpect';
import { updateSchema } from '../utils';
import * as fs from 'fs-extra';

import { getCLIPath, isCI } from '../utils';

function getSchemaPath(schemaName: string): string {
  return `${__dirname}/../../schemas/${schemaName}`;
}

export function addApiWithoutSchema(cwd: string, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true, verbose })
      .wait('Please select from one of the below mentioned services:')
      .sendline('\r')
      .wait('Provide API name:')
      .sendline('\r')
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendline('\r')
      .wait(/.*Enter a description for the API key.*/)
      .sendline('\r')
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendline('\r')
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendline('\r')
      .wait('Do you have an annotated GraphQL schema?')
      .sendline('n')
      .wait('Do you want a guided schema creation')
      .sendline('y')
      .wait('What best describes your project')
      .sendline('\r')
      .wait('Do you want to edit the schema now?')
      .sendline('n')
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud'
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function addApiWithSchema(cwd: string, schemaFile: string, verbose: boolean = !isCI()) {
  const schemaPath = getSchemaPath(schemaFile);
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true, verbose })
      .wait('Please select from one of the below mentioned services:')
      .sendline('\r')
      .wait('Provide API name:')
      .sendline('\r')
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendline('\r')
      .wait(/.*Enter a description for the API key.*/)
      .sendline('\r')
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendline('\r')
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendline('\r')
      .wait('Do you have an annotated GraphQL schema?')
      .sendline('y')
      .wait('Provide your schema file path:')
      .sendline(schemaPath)
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud'
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function addApiWithSchemaAndConflictDetection(cwd: string, schemaFile: string, verbose: boolean = !isCI()) {
  const schemaPath = getSchemaPath(schemaFile);
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true, verbose })
      .wait('Please select from one of the below mentioned services:')
      .sendline('\r')
      .wait('Provide API name:')
      .sendline('\r')
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendline('\r')
      .wait(/.*Enter a description for the API key.*/)
      .sendline('\r')
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendline('\r')
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendline('\x1b[B') // Down
      .wait(/.*Configure additional auth types.*/)
      .sendline('n')
      .wait(/.*Configure conflict detection.*/)
      .sendline('y')
      .wait(/.*Select the default resolution strategy.*/)
      .sendline('\r')
      .wait(/.*Do you have an annotated GraphQL schema.*/)
      .sendline('y')
      .wait('Provide your schema file path:')
      .sendline(schemaPath)
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud'
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function updateApiSchema(cwd: string, projectName: string, schemaName: string) {
  const testSchemaPath = getSchemaPath(schemaName);
  const schemaText = fs.readFileSync(testSchemaPath).toString();
  updateSchema(cwd, projectName, schemaText);
}

export function updateApiWithMultiAuth(cwd: string, settings: any, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['update', 'api'], { cwd, stripColors: true, verbose })
      .wait('Please select from one of the below mentioned services:')
      .sendline('')
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendline('')
      .wait(/.*Enter a description for the API key.*/)
      .sendline('description')
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendline('300')
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendline('\x1b[B') // Down
      .wait(/.*Configure additional auth types.*/)
      .sendline('y')
      .wait(/.*Choose the additional authorization types you want to configure for the API.*/)
      .sendline('a\r') // All items
      // Cognito
      .wait(/.*Do you want to use the default authentication and security configuration.*/)
      .sendline('')
      .wait('How do you want users to be able to sign in?')
      .sendline('')
      .wait('Do you want to configure advanced settings?')
      .sendline('')
      // OIDC
      .wait(/.*Enter a name for the OpenID Connect provider:.*/)
      .sendline('myoidcprovider')
      .wait(/.*Enter the OpenID Connect provider domain \(Issuer URL\).*/)
      .sendline('https://facebook.com/')
      .wait(/.*Enter the Client Id from your OpenID Client Connect application.*/)
      .sendline('clientId')
      .wait(/.*Enter the number of milliseconds a token is valid after being issued to a user.*/)
      .sendline('1000')
      .wait(/.*Enter the number of milliseconds a token is valid after being authenticated.*/)
      .sendline('2000')
      .wait('Configure conflict detection?')
      .sendline('n')
      .wait(/.*Successfully updated resource.*/)
      .sendEof()
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function updateAPIWithResolutionStrategy(cwd: string, settings: any, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['update', 'api'], { cwd, stripColors: true, verbose })
      .wait('Please select from one of the below mentioned services:')
      .sendline('')
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendline('\r')
      .wait(/.*Enter a description for the API key.*/)
      .sendline('\r')
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendline('\r')
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendline('\x1b[B') // Down
      .wait(/.*Configure additional auth types.*/)
      .sendline('n')
      .wait(/.*Configure conflict detection.*/)
      .sendline('y')
      .wait(/.*Select the default resolution strategy.*/)
      .sendline('\x1b[B') // Down
      .wait(/.*Do you want to override default per model settings.*/)
      .sendline('n')
      .wait(/.*Successfully updated resource.*/)
      .sendEof()
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}
