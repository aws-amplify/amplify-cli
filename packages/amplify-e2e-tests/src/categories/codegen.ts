import { nspawn as spawn } from '../utils/nexpect';
import { getCLIPath, isCI } from '../utils';

export function generateModels(cwd: string, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    spawn(getCLIPath(), ['codegen', 'models'], { cwd, stripColors: true, verbose }).run((err: Error) => {
      if (!err) {
        resolve();
      } else {
        reject(err);
      }
    });
  });
}
