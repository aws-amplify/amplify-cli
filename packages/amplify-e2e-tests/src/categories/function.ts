import * as nexpect from 'nexpect';
import { getCLIPath, isCI } from '../utils';

export function addHelloWorldFunction(cwd: string, settings: any, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['add', 'function'], { cwd, stripColors: true, verbose })
      .wait('Provide a friendly name for your resource to be used as a label')
      .sendline('\r')
      .wait('Provide the AWS Lambda function name')
      .sendline('\r')
      .wait('Choose the function template that you want to use')
      .sendline('\r')
      .wait('Do you want to access other resources created in this project')
      .sendline('n')
      .wait('Do you want to edit the local lambda function now')
      .sendline('n')
      .sendEof()
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function functionBuild(cwd: string, settings: any, verbose: boolean = !isCI()) {
  return new Promise((resolve, reject) => {
    nexpect
      .spawn(getCLIPath(), ['function', 'build'], { cwd, stripColors: true, verbose })
      .wait('Are you sure you want to continue building the resources?')
      .sendline('Y')
      .sendEof()
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}
