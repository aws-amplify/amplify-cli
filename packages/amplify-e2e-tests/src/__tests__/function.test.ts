import { initJSProjectWithProfile, deleteProject, amplifyPushAuth } from '../init';
import { addHelloWorldFunction, functionBuild } from '../categories/function';
import { createNewProjectDir, deleteProjectDir, getProjectMeta, getFunction } from '../utils';

describe('amplify add function', () => {
  let projRoot: string;
  beforeEach(() => {
    projRoot = createNewProjectDir();
  });

  afterEach(async () => {
    await deleteProject(projRoot);
    deleteProjectDir(projRoot);
  });

  it('init a project and add  simple function', async () => {
    await initJSProjectWithProfile(projRoot, {});
    await addHelloWorldFunction(projRoot, {});
    await functionBuild(projRoot, {});
    await amplifyPushAuth(projRoot);
    const meta = getProjectMeta(projRoot);
    const { Arn: functionArn, Name: functionName, Region: region } = Object.keys(meta.function).map(key => meta.function[key])[0].output;
    expect(functionArn).toBeDefined();
    expect(functionName).toBeDefined();
    expect(region).toBeDefined();
    const cloudFunction = await getFunction(functionName, region);
    expect(cloudFunction.Configuration.FunctionArn).toEqual(functionArn);
  });
});
