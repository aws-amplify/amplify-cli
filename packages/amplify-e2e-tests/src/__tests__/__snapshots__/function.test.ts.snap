// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodejs add function with layers can add layers configured in the project 1`] = `
Array [
  Object {
    "Fn::Sub": "arn:aws:lambda:\${AWS::Region}:\${AWS::AccountId}:layer:test-layer:1",
  },
  "arn:aws:lambda:us-west-2:123456789012:layer:my-layer:2",
]
`;

exports[`nodejs amplify add function with additional permissions environment vars comment should update on permission update 1`] = `
"/* Amplify Params - DO NOT EDIT
	ENV
	REGION
	STORAGE_NODETESTDDB_ARN
	STORAGE_NODETESTDDB_NAME
Amplify Params - DO NOT EDIT */

exports.handler = async (event) => {
    // TODO implement
    const response = {
        statusCode: 200,
        body: JSON.stringify('Hello from Lambda!'),
    };
    return response;
};
"
`;
